File: practice2/advancedHelloWorld.swift
import Foundation

let database = "totallySecureUserData.txt"
let bread = "bread"

func loadUsers() -> (users: [String: Int], lastUser: String) {
    var userData = [String: Int]()
    var lastUser = "nobody"
    
    if let contents = try? String(contentsOfFile: database) {
        let lines = contents.split(separator: "\n")
        for line in lines {
            let parts = line.split(separator: " ")
            if parts.count == 2, let count = Int(parts[1]) {
                let name = String(parts[0])
                userData[name] = count
                lastUser = name
            }
        }
    }
    
    return (userData, lastUser)
}

func saveUsers(userData: [String: Int]) {
    var output = ""
    for (name, count) in userData {
        output += "\(name) \(count)\n"
    }
    try? output.write(toFile: database, atomically: true, encoding: .utf8)
}

func handleUser(name: String, command: String) {
    var (userData, lastUser) = loadUsers()
    print("Hi there, \(lastUser)")

    if name == bread {
        try? "".write(toFile: database, atomically: true, encoding: .utf8)
        print("Everything turns into bread.")
        return
    }

    if command == "delete" {
        userData.removeValue(forKey: name)
        print("Data for \(name) have been reset.")
    } else {
        let count = userData[name] ?? 0
        if count == 0 {
            print("Welcome, \(name)!")
            userData[name] = 1
        } else {
            userData[name] = count + 1
            print("Hello again(x\(count + 1)), \(name)!")
        }
    }

    saveUsers(userData: userData)
}

func main() {
    let arguments = CommandLine.arguments
    
    if arguments.count == 1 {
        let (_, lastUser) = loadUsers()
        if lastUser != "nobody" {
            handleUser(name: lastUser, command: "")
        } else {
            print("Error: Please provide a name.")
        }
        return
    }

    if arguments.count < 2 || arguments.count > 3 {
        print("Usage: \(arguments[0]) <name> [delete]")
        return
    }

    let name = arguments[1]
    let command = arguments.count == 3 ? arguments[2] : ""
    
    handleUser(name: name, command: command)
}

main()

File: practice2/advancedHelloWorld.cpp
#include <iostream>
#include <fstream>
#include <unordered_map>
#include <string>

const std::string database = "totallySecureUserData.txt";
const std::string bread = "bread";

std::string loadUsers(std::unordered_map<std::string, int>& userData) {
    std::ifstream inFile(database);
    std::string lastUser;
    if (inFile) {
        std::string name;
        int count;
        while (inFile >> name >> count) {
            userData[name] = count;
            lastUser = name;
        }
        inFile.close();
    }
	else {
		lastUser = "nobody";
	}
    return lastUser;
}

void saveUsers(const std::unordered_map<std::string, int>& userData) {
    std::ofstream outFile(database);
    for (const auto& entry : userData) {
        outFile << entry.first << " " << entry.second << "\n";
    }
    outFile.close();
}

void handleUser(const std::string& name, const std::string& command) {
    std::unordered_map<std::string, int> userData;
    std::string lastUser = loadUsers(userData);
	std::cout << "Hi there, " << lastUser << std::endl;

    if (name == bread) {
        std::ofstream outFile(database, std::ofstream::trunc);
        outFile.close();
        std::cout << "Everything turns into bread." << std::endl;
        return;
    }

    if (command == "delete") {
        userData.erase(name);
        std::cout << "Data for " << name << " have been reset." << std::endl;
    } else {
        int& count = userData[name];
        if (count == 0) {
            std::cout << "Welcome, " << name << "!" << std::endl;
            count = 1;
        } else {
            count++;
            std::cout << "Hello again(x" << count << "), " << name << "!" << std::endl;
        }
    }

    saveUsers(userData);
}

int main(int argc, char* argv[]) {
    if (argc == 1) {
        std::unordered_map<std::string, int> userData;
        std::string lastUser = loadUsers(userData);
        if (!lastUser.empty()) {
            handleUser(lastUser, "");
        } else {
            std::cerr << "Error: Please provide a name." << std::endl;
        }
        return 0;
    }

    if (argc < 2 || argc > 3) {
        std::cerr << "Usage: " << argv[0] << " <name> [delete]" << std::endl;
        return 1;
    }

    std::string name = argv[1];
    std::string command = (argc == 3) ? argv[2] : "";

    handleUser(name, command);

    return 0;
}
